File: 0234-palindrome-linked-list/0234-palindrome-linked-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {boolean}
 */

function reverse(head)
{
    if(!head || !head.next) return head;
    
    let node = reverse(head.next);
    head.next.next = head;
    head.next = null;
    
    return node;
}

var isPalindrome = function(head) 
{
    if(!head || !head.next) return true;
    
    let p1 = head,p2 = head.next;
    
    while(p2 && p2.next)
    {
        p1 = p1.next;
        p2 = p2.next.next;
    }
    
    p2 = p1.next;
    p1.next = null;
    
    p2 = reverse(p2);
    p1 = head;
    while(p1 && p2)
    {
         if(p1.val!=p2.val) return false;
        p1 = p1.next;
        p2 = p2.next;
    }
    return true;
    
};

File: 0234-palindrome-linked-list/0234-palindrome-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head)
    {
        stack<int>st;
        
        ListNode *p1 = head;
        ListNode *p2 = head;
        
        while(p2!=NULL && p2->next!=NULL)
        {
            st.push(p1->val);
            
            p1 = p1->next;
            p2 = p2->next->next;
        }
        
        if(p2!=NULL) p1 = p1->next;
        
        while(st.size())
        {
            if(st.top()==p1->val)
            {
                st.pop();
                p1 = p1->next;
            }
            else 
                return false;
        }
        
        return true;
    }
};